sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs

```diff
@@ -0,0 +1,100 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class CollectionsShouldImplementGenericInterface : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3909";
+        private const string MessageFormat = "Refactor this collection to implement '{0}'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly Dictionary<KnownType, KnownType> nongenericToGenericMapping =
+            new Dictionary<KnownType, KnownType>
+            {
+                [KnownType.System_Collections_ICollection] = KnownType.System_Collections_Generic_ICollection_T,
+                [KnownType.System_Collections_IList] = KnownType.System_Collections_Generic_IList_T,
+                [KnownType.System_Collections_IEnumerable] = KnownType.System_Collections_Generic_IEnumerable_T,
+                [KnownType.System_Collections_CollectionBase] = KnownType.System_Collections_ObjectModel_Collection_T,
+            };
+
+        private static readonly ISet<KnownType> nongenericTypes = nongenericToGenericMapping.Keys.ToHashSet();
+        private static readonly ISet<KnownType> genericTypes = nongenericToGenericMapping.Values.ToHashSet();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+            c =>
+            {
+                var classDeclaration = c.Node as ClassDeclarationSyntax;
+
+                var issues = new List<Diagnostic>();
+
+                var implementedTypes = classDeclaration?.BaseList?.Types;
+                if (implementedTypes == null)
+                {
+                    return;
+                }
+
+                foreach (var typeSyntax in implementedTypes)
+                {
+                    var typeSymbol = c.SemanticModel.GetSymbolInfo(typeSyntax.Type).Symbol?.GetSymbolType();
+
+                    var originalDefinition = typeSymbol?.OriginalDefinition;
+                    if (originalDefinition.IsAny(genericTypes))
```

Comment:
It took me some time to understand that if we have at least one generic interface the rule does not raise issues (I thought there is a bug and I had to check the tests). Perhaps it will be more obvious if you extract the two lines above into a method `IsGenericCollection(typeSymbol)`, WDYT?